pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator.go

```
@@ -0,0 +1,102 @@
+)
+
+type VMsMutator struct {
+	ClusterConfig *virtconfig.ClusterConfig
+}
+
+func (mutator *VMsMutator) Mutate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	if !webhooks.ValidateRequestResource(ar.Request.Resource, webhooks.VirtualMachineGroupVersionResource.Group, webhooks.VirtualMachineGroupVersionResource.Resource) {
+		err := fmt.Errorf("expect resource to be '%s'", webhooks.VirtualMachineGroupVersionResource.Resource)
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	if resp := webhooks.ValidateSchema(v1.VirtualMachineGroupVersionKind, ar.Request.Object.Raw); resp != nil {
+		return resp
+	}
+
+	raw := ar.Request.Object.Raw
+	vm := v1.VirtualMachine{}
+
+	err := json.Unmarshal(raw, &vm)
+	if err != nil {
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	// Set VM defaults
+	log.Log.Object(&vm).V(4).Info("Apply defaults")
+	mutator.setDefaultMachineType(&vm)
+
+	var patch []patchOperation
+	var value interface{}
+	value = vm.Spec
+	patch = append(patch, patchOperation{
+		Op:    "replace",
+		Path:  "/spec",
+		Value: value,
+	})
+
+	value = vm.ObjectMeta
+	patch = append(patch, patchOperation{
+		Op:    "replace",
+		Path:  "/metadata",
+		Value: value,
+	})
+
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	jsonPatchType := v1beta1.PatchTypeJSONPatch
+	return &v1beta1.AdmissionResponse{
+		Allowed:   true,
+		Patch:     patchBytes,
+		PatchType: &jsonPatchType,
+	}
+}
+
+/*
+ * We're not aware of any legitimate reason to set the machine type to "".
+ * You always need to pass some machine type to qemu when starting the VM, that's the reason why we set it
+ * in the VMi mutator.
+ * Hence, we set the value also on update, so setting the machine type to "" actually means
+ * """reset to kubevirt defaults""".
+ */
+func (mutator *VMsMutator) setDefaultMachineType(vm *v1.VirtualMachine) {
+	if vm.Spec.Template == nil {
```

This would now hide cases where the validator would report that people did not specify the template. I would suggest to just do nothing, if no template is present, and let the validator fail afterwards.

---

fair enough, changing